File: java/src/main/java/world/md2html/plugins/ReplacePlugin.java
package world.md2html.plugins;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectReader;
import lombok.AllArgsConstructor;
import org.apache.commons.lang3.StringUtils;
import world.md2html.options.argfile.ArgFileParseException;
import world.md2html.options.model.Document;
import world.md2html.options.model.SessionOptions;
import world.md2html.pagemetadata.PageMetadataHandlersWrapper;
import world.md2html.utils.UserError;
import world.md2html.utils.VariableReplacer;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import static world.md2html.plugins.PluginUtils.listFromStringOrArray;
import static world.md2html.utils.JsonUtils.OBJECT_MAPPER;
import static world.md2html.utils.JsonUtils.deJson;

public class ReplacePlugin extends AbstractMd2HtmlPlugin implements PageMetadataHandler {

    @AllArgsConstructor
    private static class Replacement {
        public VariableReplacer replacer;
        public boolean recursive;
    }

    private final List<PageMetadataHandlerInfo> pageLinksHandlers = new ArrayList<>();
    private final Map<String, Replacement> replacements = new HashMap<>();
    private PageMetadataHandlersWrapper metadataHandlers;

    @Override
    public void acceptData(JsonNode data) throws ArgFileParseException {
        assureAcceptDataOnce();
        validateInputDataAgainstSchemaFromResource(data, "plugins/replace_schema.json");

        for (JsonNode node : data) {
            JsonNode markersNode = node.get("markers");
            ObjectReader reader = OBJECT_MAPPER.readerFor(new TypeReference<List<String>>() {
            });
            ArrayList<String> markers;
            try {
                markers = reader.readValue(markersNode);
            } catch (IOException e) {
                throw new ArgFileParseException("Error reading plugin '" +
                        this.getClass().getSimpleName() + "' data: " + node);
            }
            String replaceWith = node.get("replace-with").asText();
            VariableReplacer replacer;
            for (String marker : markers) {
                try {
                    replacer = new VariableReplacer(replaceWith);
                } catch (VariableReplacer.VariableReplacerException e) {
                    throw new UserError(e.getMessage() + " The template is: " + replaceWith);
                }
                boolean recursive = false;
                JsonNode recursiveNode = node.get("recursive");
                if (recursiveNode != null) {
                    recursive = recursiveNode.asBoolean(false);
                }
                replacements.put(marker.toUpperCase(), new Replacement(replacer, recursive));
            }
            this.pageLinksHandlers.addAll(markers.stream()
                    .map(m -> new PageMetadataHandlerInfo(this, m, false))
                    .collect(Collectors.toList()));
        }
    }

    @Override
    public boolean isBlank() {
        return this.pageLinksHandlers.isEmpty();
    }

    @Override
    public void acceptAppData(SessionOptions options, List<Md2HtmlPlugin> plugins,
                              PageMetadataHandlersWrapper metadataHandlers) {
        this.metadataHandlers = metadataHandlers;
    }

    @Override
    public List<PageMetadataHandlerInfo> pageMetadataHandlers() {
        return this.pageLinksHandlers;
    }

    @Override
    public String acceptPageMetadata(Document document, String marker, String metadata,
                                     String metadataSection, Set<String> visitedMarkers
    ) throws PageMetadataException {

        String metadataStr = StringUtils.stripStart(metadata, null);

        List<String> values;
        if (metadataStr.startsWith("[")) {
            //noinspection unchecked
            values = (List<String>) deJson(listFromStringOrArray(document, metadataStr));
        } else {
            values = Collections.singletonList(metadataStr);
        }
        Replacement replacement = replacements.get(marker.toUpperCase());
        String result = replacement.replacer.replace(values);

        return replacement.recursive ?
                metadataHandlers.applyMetadataHandlers(result, document, visitedMarkers) :
                result;
    }
}

File: python/plugins/replace_plugin.py
from pathlib import Path
from typing import Union, Dict

from models.document import Document
from models.options import Options
from models.page_metadata_handlers import PageMetadataHandlers
from page_metadata_utils import apply_metadata_handlers
from plugins.md2html_plugin import Md2HtmlPlugin
from plugins.plugin_utils import list_from_string_or_array
from utils import UserError, VariableReplacer

MODULE_DIR = Path(__file__).resolve().parent


class ReplacePlugin(Md2HtmlPlugin):

    def __init__(self):
        super().__init__()
        self.metadata_handlers = []
        self.all_metadata_handlers = None
        self.replacers = {}

    def accept_data(self, data):
        self.assure_accept_data_once()
        self.validate_data_with_file(data, MODULE_DIR.joinpath('replace_schema.json'))

        for datum in data:
            markers = datum.get("markers")
            replace_with = datum.get("replace-with")
            try:
                replacer = VariableReplacer(replace_with)
            except UserError as e:
                raise UserError(f"{str(e)}, the template is: '{replace_with}'")

            recursive = datum.get("recursive", False)

            self.replacers.update({k.upper(): (replacer, recursive) for k in markers})

            self.metadata_handlers.extend((self, m, False) for m in markers)

    def is_blank(self) -> bool:
        return not bool(self.metadata_handlers)

    def accept_app_data(self, plugins: list, options: Options, metadata_handlers: PageMetadataHandlers):
        self.all_metadata_handlers = metadata_handlers

    def page_metadata_handlers(self):
        return self.metadata_handlers

    def accept_page_metadata(self, doc: Document, marker: str, metadata_str: str,
                             metadata_section: str,
                             visited_markers: Union[Dict[str, None]] = None):
        # Preserving trailing spaces
        metadata_str = metadata_str.lstrip()
        try:
            metadata = list_from_string_or_array(metadata_str)
        except UserError as e:
            raise UserError(f"Error in replace entry: {str(e)}")

        replacer, recursive = self.replacers[marker.upper()]
        result = replacer.replace(metadata)
        if recursive:
            result = apply_metadata_handlers(result, self.all_metadata_handlers, doc,
                                             visited_markers=visited_markers)

        return result
